src/HtmlReader/architecture.md

```diff
@@ -0,0 +1,78 @@
+
+### Inter-resource Navigation
+
+This might be the most challenging aspect of the web reader architecture. In the Readium Architecture, a `Locator` indicates a specific location. `locator.href` is the resource, and `locator.locations` is an object containing multiple references to the specific location within the resource. It can contain:
+
+- `progression`: A decimal value representing percent of the way through the resource. This is what we typically use when navigating to scroll positions. We calculate the `progression` of the desired position and scroll the user there.
+- `fragment`: A string representing a location. Currently, we only support anchors (eg: `"#some-heading"`). We can scroll the user there by querying the dom for that element and then calling `el.scrollIntoView()`.
+- `position`: An integer representation of the location, which I interpret to mean "page number". This is not very portable because the number of pages and size of pages changes when browser size changes or font size changes. Thus, we prefer to use `progression`, but we keep `position` updated in case we wish to display it at some point.
+
+Note on CFIs: In the future, we would like to support CFIs. These are strings representing specific locations in an HTML document. They are portable as well. When you're reading a kindle and it says "location 441/2387", those numbers are the CFI that you are on and the total number of CFIs. More work on this is needed in the future.
+
+## Other effects
+
+1. `useUpdateScroll`: When the user scrolls freely, we wait for them to stop scrolling and then update dispatch an action to update the `state.location.locations.progression`, keeping the location up to date.
+1. `useLocationQuery`: We now keep a copy of `state.location` sync'd to the `?location` query param in the URL. This allws users to share or save links to specific locations. So we use an effect to update the query whenever `state.location` changes.
+1. `useWindowResize`: When the window resizes, we need to re-scroll the user to the location they are meant to be on. Thus we need to move backwards to the `NavigatingState`.
+1. `useIframeLinkClick`: Uses `React.useEffect` to add a event listener for a "message" event, which is sent from the iframe upon internal link clicks. We then dispatch a `GO_TO_HREF` with the clicked value.
+
+## Decisions
+
+- Store all state in `useReducer`.
+- Use the `webpubManifestUrl` as the base url when constructing urls from relative urls. If we get a webpub manifest with relative urls, we must assume they are relative to the webpub manifest url itself. We always resolve urls to full URL objects instead of leaving them as possible relative strings.
```

Comment:
This is great that we made this explicit! This makes things a lot easier (i.e. when comparing internal links. 😆 )

Reply:
Yea that was a pain previously!